Skip Loginview create when credentials already exist

qcserestipy opened the pull request:
Fixes #267 

# Check for Existing Credentials and Bypass Login View Creation

1. Implemented a verification step to detect existing Harbor credentials in the current configuration.
2. If valid credentials are found, the creation of a new login view is skipped.
3. Enhances login efficiency by reusing stored credentials and reducing unnecessary prompts.

ToDo: Add base64 encoding for password in config file?

qcserestipy commented:
To resolve the failing tests I will introduce a mock Keyring since the container has no keyring.

bupd reviewed (commented):
There should be an easy way to change the login credentials directly. 

- Changing login credentials should not need users to change the config.yaml

qcserestipy commented:
@bupd Do you thing it would come in handy to create a config subcommand that lets us manipulate the config file?
Such that a user could do something like `harbor config set Credential.Password <password>` to update a field in the config file.

qcserestipy commented:
@bupd  I added the config sub command and several tests for its functionality. I added some screenshots:

### Listing Config Items and Setting them
<img width="839" alt="Screenshot 2024-12-22 at 13 53 59" src="https://github.com/user-attachments/assets/28fb0bd1-5c66-48cf-8626-f80888c57345" />

### Clearing config items
<img width="655" alt="Screenshot 2024-12-22 at 13 59 58" src="https://github.com/user-attachments/assets/5558fb58-6527-46f3-8343-4fdeaa8ad226" />

### Retrieving config items
<img width="636" alt="Screenshot 2024-12-22 at 13 56 46" src="https://github.com/user-attachments/assets/f8b59781-b880-4e67-9873-f875721b193f" />

### Setting credentials and credentialname supported with --name flag
<img width="1002" alt="Screenshot 2024-12-22 at 13 56 01" src="https://github.com/user-attachments/assets/84cb5735-8e8f-4716-b1f9-36401e399cea" />

<img width="972" alt="Screenshot 2024-12-22 at 13 56 11" src="https://github.com/user-attachments/assets/ed43d5a9-7bae-48aa-b958-cf98a380c1c0" />

<img width="945" alt="Screenshot 2024-12-22 at 13 54 25" src="https://github.com/user-attachments/assets/c2f787c1-5390-402d-ad77-86e1a82d0180" />

qcserestipy commented:
@bupd There is still a sign off error in one of the commits. Unfortunately, I would have to rebase over 34 commits from main. Is there an easier way to do this?

bupd commented:
Try squashing the commits

bupd commented:
@qcserestipy did you also add changing between different login accounts via changing the credential name command. 

Which utilizes the current credential name.

qcserestipy commented:
@bupd I went through the rebase and it worked with a little bit of manual work.
Concerning your second question can you give me an example what you mean?

You can set the `current-credential-name` with the command too:

<img width="746" alt="Screenshot 2024-12-22 at 14 33 19" src="https://github.com/user-attachments/assets/a8d4dc67-4ade-4469-bc5b-ec00df01ab97" />

Standing-Man commented:
Hi @qcserestipy and @bupd, If the command format is as follows, it might be even better.
- `harbor context list` -> list all context
- `harbor context create` -> create a new context for now 
- `harbor context view` -> view current context
-  `harbor context delete` -> delete context 
- ....

This is just a suggestion I have for this command. I'm not sure if it would be more user-friendly, but I’d love to hear your thoughts!

bupd reviewed (changes requested):
There is some weird behaviour for me while testing this.
1. config file gets resurrected automagically with previous credentials.

Attaching video for reference.

https://github.com/user-attachments/assets/73ffbf35-ccdd-460d-97b1-88cf581ef8a9


2. I can't create a new login.
3. Main focus of config command should be - allowing users to change the current credential or login with new ones. Since people should have ability to login with different harbor instances, directly from the CLI

bupd commented:
@qcserestipy to reduce the complexity. 
I would suggest using the login command to switch between contexts and also offer. to use new login too.

qcserestipy commented:
@bupd I made the changes that you suggested. Users can now change the context via the `--name` flag of the login command.

Vad1mo commented on doc/man-docs/man1/harbor-config-set.1:
```diff
@@ -0,0 +1,57 @@
+  harbor config set credentials.password myNewSecret
+
+  # Set a credential's password by specifying the credential name
+  harbor config set credentials.password myNewSecret --name harbor-cli@http://demo.goharbor.io
```
the example should be a bit different here. eg. 

```
credentials:
    - name: admin@demo-goharbor-io

```

Vad1mo reviewed (commented):
deleting the whole context entry woudl be nice too.

Vad1mo commented:
We did some testing and in generally it looks good to me.

bupd commented on cmd/harbor/root/config/delete.go:
```diff
@@ -0,0 +1,244 @@
+  # Clear a specific credential's password using --name
+  harbor config delete credentials.password --name admin@http://demo.goharbor.io
+
+  # Clear all credentials and the current credential
+  harbor config delete --all
```
can we make  the flag `--current` instead of `--all` since all would mean delete all of the credentials.
```suggestion
  # Clear the current credential
  harbor config delete --all
```

bupd replied:
BTW can we just set to delete the current config with both username and password by default.

qcserestipy replied:
What do you mean with `by default`? It is implemented such that the `--current` flag deletes the whole credentials list item. Is that what you mean or can you give me an example?

bupd commented on cmd/harbor/root/config/update.go:
```diff
@@ -0,0 +1,267 @@
+	var credentialName string
+
+	cmd := &cobra.Command{
+		Use:   "set <item> <value>",
```
Instead of set we should call it update here.

```suggestion
		Use:   "update <item> <value>",
```

bupd reviewed (changes requested):
Thanks for the contributions 

Left some minor nitpicks.

bupd commented:
Overall Looks Good to me. But the only thing stopping us from merging is the git commit history of this PR. which includes duplicate commits which is already in main branch.

If you are done with the fixes let me know. I am ready to help. We can do a cherry-pick and create squashed new PR only with necessary commits. 

Hope this helps. Thanks

qcserestipy commented:
@bupd Thank you for the extensive reviews! I am happy to hear this code is nearly ready for being merged. From my side I would be ready for making a squashed PR with necessary commits.

qcserestipy commented:
Since #372 was merged we can close this one.
